src/cache/local.ml

```diff
@@ -72,92 +72,60 @@ let detect_unexpected_dirs_under_cache_root cache =
-
-   CR-someday amokhov: In Dune we generally assume that digest collisions are
-   impossible, so it seems better to remove this logic in future. *)
-module Collision_chain = struct
-  type search_result =
-    | Found of { existing_path : Path.t }
-    | Not_found of { next_available_path : Path.t }
-
-  (* This function assumes that we do not create holes in the suffix numbering. *)
-  let search path file =
-    let rec loop n =
-      let path = Path.extend_basename path ~suffix:("." ^ string_of_int n) in
-      if Path.exists path then
-        if Io.compare_files path file = Ordering.Eq then
-          Found { existing_path = path }
-        else
-          loop (n + 1)
-      else
-        Not_found { next_available_path = path }
-    in
-    loop 1
-end
-
 (* A file storage scheme. *)
 module type FSScheme = sig
   (* Given a cache root and a file digest, determine the location of the file in
      the cache. *)
   val path : root:Path.t -> Digest.t -> Path.t
 
-  (* Extract a file's digest from its location in the cache. *)
-  val digest : Path.t -> Digest.t
-
   (* Given a cache root, list all files stored in the cache. *)
   val list : root:Path.t -> Path.t list
 end
 
 (* A file storage scheme where a file with a digest [d] is stored in a
    subdirectory whose name is made of the first two characters of [d], that is:
 
-   [<root>/<first-two-characters-of-d>/<d>.<N>]
-
-   The suffix [.<N>] is used to handle collisions, i.e. the (unlikely)
-   situations where two files have the same digest.
+   [<root>/<first-two-characters-of-d>/<d>]
 
-   CR-soon amokhov: Note that the function [path] returns the path without the
-   [.<N>] suffix, whereas the function [digest] expects the [.<N>] suffix to be
-   present. We should fix this inconsistency. *)
+   CR-someday amokhov: We currently do not provide support for collisions where
+   two or more files with different content have the same content digest [d]. If
+   this ever becomes a real (i.e. not hypothetical) problem, a good way forward
+   would be to switch from MD5 to SHA1 or higher, making the chance of this
+   happening even lower, and at the same time making it more difficult to create
+   deliberate collisions. *)
 module FirstTwoCharsSubdir : FSScheme = struct
   let path ~root digest =
     let digest = Digest.to_string digest in
     let first_two_chars = String.sub digest ~pos:0 ~len:2 in
     Path.L.relative root [ first_two_chars; digest ]
 
-  let digest path =
-    match Digest.from_hex (Path.basename (fst (Path.split_extension path))) with
-    | Some digest -> digest
-    | None ->
-      Code_error.raise "strange cached file path (not a valid digest)"
-        [ (Path.to_string path, Path.to_dyn path) ]
-
   let list ~root =
+    let open Result.O in
     let f dir =
-      let is_hex_char c =
-        let char_in s e = Char.compare c s >= 0 && Char.compare c e <= 0 in
-        char_in 'a' 'f' || char_in '0' '9'
+      let is_hex_char = function
```

Comment:
Would be nice to have a `Char` module in Stdune with this function.

Reply:
Yep, makes sense, I'll do this in a separate PR.